src/backend/distributed/master/master_multi_shard_modify.c

```diff
@@ -0,0 +1,402 @@
+#include "distributed/multi_client_executor.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_router_executor.h"
+#include "distributed/multi_router_planner.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/multi_transaction.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/resource_lock.h"
+#include "distributed/worker_protocol.h"
+#include "optimizer/clauses.h"
+#include "optimizer/predtest.h"
+#include "optimizer/restrictinfo.h"
+#include "optimizer/var.h"
+#include "nodes/makefuncs.h"
+#include "tcop/tcopprot.h"
+#include "utils/builtins.h"
+#include "utils/datum.h"
+#include "utils/inval.h"
+#include "utils/lsyscache.h"
+
+
+static int SendQueryToShards(Query *query, List *shardIntervalList);
+static HTAB * OpenConnectionsToAllShardPlacements(List *shardIntervalList);
+static void OpenConnectionsToShardPlacements(uint64 shardId, HTAB *shardConnectionHash);
+
+static int SendQueryToConnections(char *shardQueryString,
+								  ShardConnections *shardConnections);
+static void LockShardsForModify(List *shardIntervalList);
+PG_FUNCTION_INFO_V1(master_multi_shard_modify);
+
+
+/*
+ * master_multi_shard_modify takes in a delete query string and pushes the
+ * query to shards. It finds shards that match the criteria defined in the
+ * delete command, generates the same delete query string for each of the
+ * found shards with distributed table name replaced with the shard name and
+ * sends the queries to the workers. It uses one-phase or two-phase commit
+ * transactions depending on citus.copy_transaction_manager value.
+ */
+Datum
+master_multi_shard_modify(PG_FUNCTION_ARGS)
+{
+	text *queryText = PG_GETARG_TEXT_P(0);
+	char *queryString = text_to_cstring(queryText);
+	List *queryTreeList = NIL;
+	Oid relationId = InvalidOid;
+	Index tableId = 1;
+	Query *modifyQuery = NULL;
+	Node *queryTreeNode;
+	List *restrictClauseList = NIL;
+	bool isTopLevel = true;
+	bool failOK = false;
+	List *shardIntervalList = NIL;
+	List *prunedShardIntervalList = NIL;
+	int32 affectedTupleCount = 0;
+
+	PreventTransactionChain(isTopLevel, "master_multi_shard_modify");
+
+	queryTreeNode = ParseTreeNode(queryString);
+	if (IsA(queryTreeNode, DeleteStmt))
+	{
+		DeleteStmt *deleteStatement = (DeleteStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(deleteStatement->relation, NoLock, failOK);
+	}
+	else if(IsA(queryTreeNode, UpdateStmt))
+	{
+		UpdateStmt *updateStatement = (UpdateStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(updateStatement->relation, NoLock, failOK);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete nor update statement",
+							   queryString)));
+	}
+
+	CheckDistributedTable(relationId);
+
+	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
+	modifyQuery = (Query *) linitial(queryTreeList);
+
+	ErrorIfModifyQueryNotSupported(modifyQuery);
```

Comment:
We have synced with Brian. This UDF will not support STABLE functions in UPDATE queries and Brian's changes in #488 allows STABLE functions with UPDATE queries in some cases. When #488 is merged, ErrorIfModifyQueryNotSupported function won't be accurate for this UDF. 

Basically, we can add a static function that basically call ErrorIfModifyQueryNotSupported function and also error out with STABLE functions. Should we include it in this PR or should we add it with #488 or after #488 as another PR?